python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+        yn_id = DynamicID(type=Constants.TYPE_OF_RESOURCE, id=vm_id)
+
+        # List all the tags present on the VM.
+        for tag_id in self.vsphere_sdk_client.tagging.TagAssociation.list_attached_tags(
+            yn_id
+        ):
+            vsphere_vm_tag = self.vsphere_sdk_client.tagging.Tag.get(tag_id=tag_id).name
+            tag_key_value = vsphere_tag_to_kv_pair(vsphere_vm_tag)
+            tag_key = tag_key_value[0] if tag_key_value else None
+            if tag_key == tag_key_to_remove:
+                # Remove the tag matching the key passed.
+                cli_logger.info("Removing tag {} from the VM {}".format(tag_key, vm_id))
+                self.vsphere_sdk_client.tagging.TagAssociation.detach(tag_id, yn_id)
+                break
+
+    def get_frozen_vm_obj(self):
+        frozen_vm_target_name = self.frozen_vm_name
+        vms = self.vsphere_sdk_client.vcenter.VM.list(
+            VM.FilterSpec(names={frozen_vm_target_name})
+        )
+
+        cli_logger.info("VM to clone from {}".format(vms))
+
+        return vms[0] if len(vms) > 0 else None
+
+    def get_pyvmomi_obj(self, vimtype, name):
+        obj = None
+
+        # TODO: Find a better way to solve pyvmomi timeout issues
+        self.vsphere_pyvmomi_sdk_client = VmwSdkClient(
+            self.vsphere_credentials["server"],
+            self.vsphere_credentials["user"],
+            self.vsphere_credentials["password"],
+            VmwSdkClient.SessionType.UNVERIFIED,
+            VmwSdkClient.ClientType.PYVMOMI_SDK,
+        ).get_client()
+
+        container = self.vsphere_pyvmomi_sdk_client.viewManager.CreateContainerView(
+            self.vsphere_pyvmomi_sdk_client.rootFolder, vimtype, True
+        )
+
+        for c in container.view:
+            if name:
+                if c.name == name:
+                    obj = c
+                    break
+            else:
+                obj = c
+                break
+
+        return obj
+
+    def connect_nics(self, vm_id):
+        nics = self.vsphere_sdk_client.vcenter.vm.hardware.Ethernet.list(vm_id)
+
+        for nic in nics:
+            eth_update_spec = Ethernet.UpdateSpec(start_connected=True)
+            self.vsphere_sdk_client.vcenter.vm.hardware.Ethernet.update(
+                vm_id, nic.nic, eth_update_spec
+            )
+            cli_logger.info("Connecting NIC {} to VM {}".format(nic, vm_id))
+            self.vsphere_sdk_client.vcenter.vm.hardware.Ethernet.connect(vm_id, nic.nic)
+
+    # This method is used to tag VMs as soon as they show up on vCenter.
+    def tag_vm(self, vm_name, tags):
+        names = {vm_name}
+
+        start = time.time()
+        # In most cases the instant clone VM will show up in several seconds.
+        # When the vCenter Server is busy, the time could be longer. We set a 120
+        # seconds timeout here. Because it's not used anywhere else, we don't make
+        # it as a formal constant.
+        timeout = 120
```

Comment:
Is it worth making this configurable via environment variable?

Reply:
I don't think we want he user to to configure this because this is our internal logic.
We do can make it a constant.

Reply:
Have change the "120" to a constant. For now we don't want it to be configurable by users.